core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -764,16 +781,24 @@ private DeleteResult innerDelete(Delete delete) throws IOException {
             }
 
             final long expectedVersion = delete.version();
-            if (checkVersionConflict(delete, currentVersion, expectedVersion, deleted)) {
-                // skip executing delete because of version conflict on recovery
-                deleteResult = new DeleteResult(expectedVersion, SequenceNumbersService.UNASSIGNED_SEQ_NO, true);
-            } else {
-                final long seqNo;
-                if (delete.origin() == Operation.Origin.PRIMARY) {
+
+            final boolean conflict = checkVersionConflict(delete, currentVersion, expectedVersion, deleted);
+
+            final long seqNo;
+            if (delete.origin() == Operation.Origin.PRIMARY) {
+                if (!conflict) {
                     seqNo = seqNoService.generateSeqNo();
                 } else {
-                    seqNo = delete.seqNo();
+                    seqNo = SequenceNumbersService.UNASSIGNED_SEQ_NO;
                 }
+            } else {
+                seqNo = delete.seqNo();
+            }
+
+            if (checkVersionConflict(delete, currentVersion, expectedVersion, deleted)) {
```

Comment:
Also - it feels like this can simplified based on the origin. I do wonder if the code reuse in `checkVersionConflict` is worth the extra complexity it brings to reading the code.